common/model/transfer.c

```diff
@@ -0,0 +1,715 @@
+  }
+  // Should not reach here
+  return NULL;
+}
+
+// Get the transfer data
+transfer_data_t transfer_data(transfer_t transfer) {
+  if (transfer_type(transfer) != DATA) {
+    return NULL;
+  }
+  return transfer->data;
+}
+
+// Get the transfer value in
+transfer_value_out_t transfer_value_out(transfer_t transfer) {
+  if (transfer_type(transfer) != VALUE_OUT) {
+    return NULL;
+  }
+  return transfer->in;
+}
+
+// Get the transfer value out
+transfer_value_in_t transfer_value_in(transfer_t transfer) {
+  if (transfer_type(transfer) != VALUE_IN) {
+    return NULL;
+  }
+  return transfer->out;
+}
+
+// Creates and returns a new transfer
+transfer_t transfer_new(transfer_type_e transfer_type) {
+  transfer_t transfer;
+  transfer = (transfer_t)malloc(sizeof(struct _transfer));
+  if (!transfer) {
+    // errno = IOTA_OUT_OF_MEMORY
+    return NULL;
+  }
+  memset(transfer, 0, sizeof(struct _transfer));
+  memset(transfer->tag, FLEX_TRIT_NULL_VALUE, sizeof(transfer->tag));
+  transfer->type = transfer_type;
+  switch (transfer_type) {
+    case DATA:
+      transfer->data = transfer_data_new();
+      break;
+    case VALUE_OUT:
+      transfer->in = transfer_value_out_new();
+      break;
+    case VALUE_IN:
+      transfer->out = transfer_value_in_new();
+      break;
+    default:
+      free(transfer);
+      transfer = NULL;
+  }
+  return transfer;
+}
+
+// Free an existing transfer
+void transfer_free(transfer_t transfer) {
+  switch (transfer_type(transfer)) {
+    case DATA:
+      transfer_data_free(transfer_data(transfer));
+      break;
+    case VALUE_OUT:
+      transfer_value_out_free(transfer_value_out(transfer));
+      break;
+    case VALUE_IN:
+      transfer_value_in_free(transfer_value_in(transfer));
+      break;
+  }
+  free(transfer);
+}
+
+/***********************************************************************************************************
+ * Transfer Context data structure
+ ***********************************************************************************************************/
+// To obfuscate the model, the structure definition comes here
+// struct _transfer_ctx {...};
+
+// Creates and returns a new transfer context
+int transfer_ctx_init(transfer_ctx_t transfer_ctx, transfer_t *transfers,
+                      size_t len) {
+  size_t i;
+  int64_t total = 0;
+  for (i = 0; i < len; i++) {
+    transfer_t transfer = transfers[i];
+    transfer_ctx->count += transfer_transactions_count(transfer);
+    total += transfer_value(transfer);
+  }
+  return total ? -1 : 0;
+}
+
+// Gets the count of transactions for the transfers
+int transfer_ctx_count(transfer_ctx_t transfer_ctx) {
+  return transfer_ctx->count;
+}
+
+// Calculates the bundle hash for a collection of transfers
+int transfer_ctx_hash(transfer_ctx_t transfer_ctx, Kerl *kerl,
+                      transfer_t *transfers, size_t len) {
+  size_t i, j, count, current_index = 0;
+  tryte_t essence[TRYTES_PER_ESSENCE];

```

Comment:
In the 4:1 and 5:1 models, we would need to use the 'insert' function which would be far less efficient. When kerl support flex_trit natively, it will become advantageous since we can remove the conversion from trytes (or flex_trits) to trits also.